components/builder-depot/src/backend/s3.rs

```diff
@@ -169,45 +193,53 @@ impl S3Handler {
         }
     }
 
-    fn single_upload(&self, key: String, hart: File, path_attr: &str) -> Result<()> {
+    fn single_upload(&self, key: &str, hart: File, path_attr: &PathBuf) -> Result<()> {
         Counter::SingleUploadRequests.increment();
         let start_time = PreciseTime::now();
         let mut reader = BufReader::new(hart);
         let mut object: Vec<u8> = Vec::new();
         let bucket = self.bucket.clone();
-        let _complete = reader.read_to_end(&mut object)?;
+        let _complete = reader.read_to_end(&mut object).map_err(Error::IO);
 
         let mut request = PutObjectRequest::default();
-        request.key = key;
+        request.key = key.to_string();
         request.bucket = bucket;
         request.body = Some(object);
 
-        match self.client.put_object(&request).sync() {
-            Ok(_) => {
-                let end_time = PreciseTime::now();
-                info!(
-                    "Upload completed for {} (in {} sec):",
-                    &path_attr,
-                    start_time.to(end_time)
-                );
-                Ok(())
-            }
-            Err(e) => {
-                Counter::UploadFailures.increment();
-                warn!("Upload failed for {}: ({:?})", &path_attr, e);
-                Err(Error::PackageUpload(e))
+        let fqpi = path_attr.clone().into_os_string().into_string().unwrap();
+
+        let put_future = self.client.put_object(&request).then(move |result| {
+            let end_time = PreciseTime::now();
+            match result {
+                Ok(_) => {
+                    info!(
+                        "Upload completed for {} (in {} sec):",
+                        fqpi,
+                        start_time.to(end_time)
+                    );
+                    return Ok(());
+                }
+                Err(e) => {
+                    Counter::UploadFailures.increment();
+                    warn!("Upload failed for {}: ({:?})", fqpi, e);
+                    return Err(Error::PackageUpload(e));
+                }
```

Comment:
It's possible I'm misreading this, but off the cuff, it looks like this might be introducing a subtle bug.  The rusoto docs say that `sync()` blocks the current thread until the future has resolved, which effectively gives you synchronous behavior.

I haven't read the docs for Rust futures, but in other languages, calling `then` on a future registers a callback that will be run whenever that future happens to complete, whether that's 5 seconds or 5 minutes from now.  It returns immediately, so the executing thread won't block.

So before, when this function returned, we were confident that all of this code had finished executing, but now, that's not the case.

Am I reading this right?

Reply:
You are correct, the `then` call chains a computation on to the returned future - the `then` combinator will get executed whether the returned future is a failure or a success.

Reply:
Then I don't think this is going to be what we want.